File: go/parsers/goodbye.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type GoodBye struct {
	gen *gen.Goodbye
}

func NewGoodByeFields(g *gen.Goodbye) messages.GoodByeFields {
	return &GoodBye{gen: g}
}

func (g *GoodBye) Reason() string {
	reason, _ := g.gen.Reason()
	return reason
}

func (g *GoodBye) Details() map[string]any {
	return map[string]any{}
}

func GoodbyeToCapnproto(m *messages.GoodBye) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	goodbye, err := gen.NewRootGoodbye(seg)
	if err != nil {
		return nil, err
	}

	if err := goodbye.SetReason(m.Reason()); err != nil {
		return nil, err
	}

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeGoodbye, data, nil), nil
}

func CapnprotoToGoodbye(data []byte) (*messages.GoodBye, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	goodbye, err := gen.ReadRootGoodbye(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewGoodByeWithFields(NewGoodByeFields(&goodbye)), nil
}

File: python/wamp_msgs_capnp/parsers/goodbye.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.goodbye import Goodbye, IGoodbyeFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "goodbye.capnp")
goodbye_capnp = capnp.load(str(module_file))


class GoodbyeFields(IGoodbyeFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def reason(self) -> str:
        return self._gen.reason

    @property
    def details(self) -> dict[str, Any]:
        return {}


def goodbye_to_capnproto(g: Goodbye) -> bytes:
    goodbye = goodbye_capnp.Goodbye.new_message()
    goodbye.reason = g.reason

    packed_data = goodbye.to_bytes_packed()

    return helpers.prepend_header(Goodbye.TYPE, packed_data, b"")


def capnproto_to_goodbye(data: bytes) -> Goodbye:
    message_data, _ = helpers.extract_message(data)
    goodbye_obj = goodbye_capnp.Goodbye.from_bytes_packed(message_data)

    return Goodbye(GoodbyeFields(goodbye_obj))
